Virtual frames.

psivaraj opened the pull request:
First stab at an OpenPulse language proposal using virtual frames.

psivaraj committed:
Virtual frames instantiation.

Co-authored-by: Peter Karalekas <peter@karalekas.com>
Co-authored-by: Steven Heidel <steven@heidel.ca>
Co-authored-by: Colm Ryan <colm@colmryan.org>

psivaraj committed:
Add `prelude` and minor refactor.

taalexander commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
 
-.. code-block:: none
+  frame driveframe1 = Frame(1.0, 5e9, 0.0);
+  frame driveframe2 = copy(driveframe1, phase=driveframe1.phase + pi/2);
```
I would tend to lean away from named arguments as these are not currently part of OpenQASM. I know we are free to do whatever we want, but I would err on the side of consistency.

psivaraj replied:
Sounds fine by me. We can just set it after the copy.

taalexander commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
+.. code-block:: javascript
 
-.. code-block:: none
+  frame driveframe1 = Frame(1.0, 5e9, 0.0);
```
Why capital? To differentiate the type and initializer?

psivaraj replied:
Yes; other suggestions welcome.

taalexander commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
 
-   // Shift phase of qubit 0's "drive" frame by pi/4, eg. an rz gate with angle -pi/4
-   frameof("drive", %0).phase += pi/4;
+This will generate a ::math:`pi/2` phase incremented copy of ``driveframe1`` (i.e. with
```
Is this independent of the derived frame from this point forward?

psivaraj replied:
No; there are no dependencies between frames.

stevenheidel replied:
Took me a few seconds to parse, I think you mean "yes; there are no dependencies between frames"

caryan replied:
The parenthetical "i.e. with the same frequency and time as driveframe1" might just make this more confusing. Just saying it's a copy might be enough.

zachschoenfeld33 replied:
Would it be beneficial to define the copied frame w/ no aggregated phase and initial time = 0? This seems to be the main reason to copy a `frame`. For instance, if you have already used the `frame` and want a fresh one to use on a new physical channel. If you wanted the same time, you could just keep using the original frame.

psivaraj replied:
> The parenthetical "i.e. with the same frequency and time as driveframe1" might just make this more confusing. Just saying it's a copy might be enough.

Fixed.

> Would it be beneficial to define the copied frame w/ no aggregated phase and initial time = 0? This seems to be the main reason to copy a frame. For instance, if you have already used the frame and want a fresh one to use on a new physical channel. If you wanted the same time, you could just keep using the original frame.

You can do that by initializing a new frame via the `Frame(freq, phase)` initialize. The copy is there for the explicit cases where you want to same time as same other frame.

psivaraj replied:
> The parenthetical "i.e. with the same frequency and time as driveframe1" might just make this more confusing. Just saying it's a copy might be enough.

Fixed.

> Would it be beneficial to define the copied frame w/ no aggregated phase and initial time = 0? This seems to be the main reason to copy a frame. For instance, if you have already used the frame and want a fresh one to use on a new physical channel. If you wanted the same time, you could just keep using the original frame.

You can do that by initializing a new frame via the `Frame(freq, phase)` initialize. The copy is there for the explicit cases where you want to same time as same other frame.

taalexander commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
-   // Equivalent
-   const drive = frameof("drive", %0);
-   drive.phase += pi/4;
+Like other things in OpenQASM, a ``frame`` is locally scoped unless a ``prelude`` defcal is used
```
Agreed, we need to decide how to do scoping.

stevenheidel replied:
I wonder if we named it `global` that would make it more clear

taalexander replied:
Yes, that may be clearer. Possibly something like 

```
calconfig {

}
```

May make sense as well. A block scoped just for setup for a calibration grammar.

psivaraj replied:
Gave it a shot with `calconfig`.

taalexander commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
-     frameof("drive", %q).phase -= theta;
+.. code-block:: javascript
+
+   defcal prelude {
```
I was wondering if we should consider getting qubit wildcarded attributes, this would allow more generic defcals that work for many qubits, eg., 

```c
defcal prelude {
     set_attr("frame",  Frame(5.0e9, 0)) $0;
}

defcal rz(angle[20] theta) $q {
     frame carrier = get_attr("frame") $q;
     carrier.phase -= theta;
}

psivaraj replied:
Took a shot at it.

taalexander commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
+     frame driveframe0 = Frame(1.0, 5e9, 0.0);
+     frame driveframe1 = Frame(1.0, 6e9, 0.0);
+     ...
+     map[int, frame] frame_mapping = {0: driveframe0, 1: driveframe1};
```
Oh, I see. This is another way of approaching this.

psivaraj replied:
Possible overextension of the language; the `set_attr` and `get_attr` seems like a legitimate alternative.

stevenheidel replied:
I like the get/set attr idea vs adding multiple container types to the language. I would also advocate for making it narrower, `getframe` and `setframe` that only work with frames and don't need to work for other types.

taalexander replied:
> getframe and setframe that only work with frames and don't need to work for other types.

This would certainly be easier and be less of a headache. The only other piece that comes to mind would be pulses (eg., applying a scaling to an Xp pulse to get an X90) / virtual channels.

psivaraj replied:
Left it `getframe` and `setframe` for now. I'm also unclear on where frequencies or calibrated phases would come from (e.g. undoing local `RZ`s from a 2Q gate).

taalexander commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
+.. code-block:: javascript
+
+   // Shift phase of the "drive" frame by pi/4, eg. an rz gate with angle -pi/4
+   driveframe.phase += pi/4;
```
What update time does this operation take place at?

psivaraj replied:
Since each `frame` is responsible for keeping its own time, the update will occur at the current time of the `driveframe`.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
+- Tracking time appropriately so programs do not need to deal in absolute time
+- Tracking phase by producing a complex value given an input time (i.e. via the mathematical
+  relationship :math:`e^{i\left(2\pi f t + \theta\right)}`,  where `f` is frequency and
+  :math:`\theta` is phase). One motivation for keeping track of phase is to allow pulses to be be
```
```suggestion
  :math:`\theta` is phase). One motivation for keeping track of phase is to allow pulses to be
```

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
-There are two important considersations when dealing with frames.
+1. A frequency ``frequency`` of type ``float``.
+2. A phase ``phase`` of type ``angle``.
+3. A time of type ``dt`` which is manipulated implicitly and cannot be modified other
```
Perhaps `dt` should be a per channel concern?

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
-   // Equivalent
-   const drive = frameof("drive", %0);
-   drive.phase += pi/4;
+Like other things in OpenQASM, a ``frame`` is locally scoped unless a ``prelude`` defcal is used
```
I like this idea

taalexander commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
+Waveforms are scheduled using the ``play`` instruction. These instructions may
 only appear inside a ``defcal`` block!
 
 Play instructions have two required parameters:
```
Is this 3 required arguments?

zachschoenfeld33 replied:
Believe this should be updated--there is now a third required parameter, ie the channel.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+The frame is composed of three parts:
 
-There are two important considersations when dealing with frames.
```
Minor, but should make the amplitude vs non amplitude consistent in the doc (sometimes two params are used, sometimes three)

psivaraj replied:
Not sure to which ambiguity you are referring.  Could you take a look at the most recent commit and see if I fixed it?

caryan replied:
I was a bit lost on what the two or three parameters to the constructor were. Here we have `Frame(5e9, 0.0)` but below we have `Frame(1.0, 5e9, 0.0)` What about explicitly telling the reader what the arguments are and relating them back to the mathematical equation above:

```
Frames are purely virtual and can be constructed using the ``Frame`` command. E.g. 
frame driveframe1 = Frame(1.0, 5e9, 0.0);
would construct a frame with scale 1.0, frequency `f`  5 GHz and phase `\theta` 0.0.
```

zachschoenfeld33 replied:
`driveframe` has two input parameters to `Frame`. `driveframe1` below has 3 input parameters. I assume the 3rd parameter refers to an amplitude. We should be consistent on whether or not the `Frame` allows an amplitude to be specified.

psivaraj replied:
> What about explicitly telling the reader what the arguments are and relating them back to the mathematical equation above:

Added

> 

driveframe has two input parameters to Frame. driveframe1 below has 3 input parameters. I assume the 3rd parameter refers to an amplitude. We should be consistent on whether or not the Frame allows an amplitude to be specified.

Yes, that's a typo and just fixed. No `scale` in a `frame`.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
+     frame driveframe0 = Frame(1.0, 5e9, 0.0);
+     frame driveframe1 = Frame(1.0, 6e9, 0.0);
+     ...
+     map[int, frame] frame_mapping = {0: driveframe0, 1: driveframe1};
```
Neat

taalexander commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
 
    // Play a 3 sample pulse on qubit 0's "drive" frame
-   play([1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j]) frameof("drive", %0);
+   play(tx0, [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j], driveframe);
```
Could we tie channels with frames in the form of virtual channels, eg., `vc chan = virtual(channel, frame)` from this point on a virtual channel is a frame+channel and all updates to the frame have a well-defined time.

zachschoenfeld33 replied:
I like this. Also could remove the need to copy frames if we then tracked time via the virtual channels.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
+     frame qframe_12 = frame_mapping[$q][1];
+   }
+
+Note that ``map`` and ``list`` are simple base containers and are not tied to the ``frame``
```
There have been discussion to add  `array` to OpenQASM--this could replace `list`. But seems like a good idea

psivaraj replied:
:fire: for now in place of `setframe` and `getframe`.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
+   driveframe.phase += pi/4;
+
+   // Define a calibration for the rz gate on all physical qubits
+   defcal rz(angle[20]:theta) %q {
```
```suggestion
   defcal rz(angle[20] theta) $q {
```

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
-   // center is the mean of pulse
-   // sigma is the standard deviation of pulse
+   // arbitrary complex samples
+   arb_waveform = [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j];
```
Note: Will need to define array and complex types either in OpenQASM or specifically at the OpenPulse level. There have been discussion for both at the OpenQASM level

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
+   // amp is waveform amplitude at center
+   // center is the mean of waveform
+   // sigma is the standard deviation of waveform
    gaussian(length:l, complex[float[32]]:amp, length:center, length:sigma)
```
We could consider removing `center` and including a `shift` function for `waveform`'s (I think `shift` is in the signal proposal, but not included in the transformation ops here).

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
    drag(length:l, complex[float[32]]:amp, length:center, length:sigma, float[32]:beta)
 
-   // Define a constant pulse of length l
+   // Define a constant waveform of length l
```
This should include an amplitude

caryan commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
 
-This has the advantage that one can run any program with any arbitrary frame
-names provided a mapping to the appropriate channels is given.
+- Tracking time appropriately so programs do not need to deal in absolute time
```
nor the bookkeeping of advancing time in a sequence of pulses

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
 ----------------
 
-Pulses are scheduled using the ``play`` instruction. These instructions may
+Waveforms are scheduled using the ``play`` instruction. These instructions may
```
Should we note that it is illegal to use a `Frame` simultaneously across channels (aka it will invoke a compilation error)?

psivaraj replied:
Simultaneously in the sense of two `defcal` blocks scheduled simultaneously and making use of the same frame?

taalexander replied:
This is not possible. Consistency is always enforced by the relative ordering of instructions and the resulting dispatch resolution chain.

```
defcal x $0 {
    frame.phase += pi/2;
}

defcal x $1 {
    play frame, pulse
}


x $1;  // pulse happens first
x $0; // frame change happens second.
```

zachschoenfeld33 replied:
No, I mean invoking play on two different physical channels with the same `Frame`. Then they'd occur at the same time which would be an issue

psivaraj replied:
> No, I mean invoking play on two different physical channels with the same Frame. 

There wouldn't be an ambiguity (aside from sample rate issues) -- the time on the Frame would be advanced based on ordering of those two "play" instructions.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
-.. code-block:: none
+.. code-block:: javascript
 
    // Play a 3 sample pulse on qubit 0's "drive" frame
```
Update comment w/ `tx0` description

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
-   play([1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j]) frameof("drive", %0);
+   play(tx0, [1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j], driveframe);
 
    // Play a gaussian on qubit 1's "drive" frame
```
Same comment on `tx0`

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -274,13 +282,13 @@ The following are possible parameters that might be included:
 Again it is up to the hardware vendor to determine the parameters and write a
```
On line 274, says "only required parameter is a `frame`". Should update to include the channel.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -290,28 +298,262 @@ pushing the results into some buffer which is then accessed outside the program.
+   defcal aligned_gates {
+     // driveframe1 and driveframe2 used in this defcal, so clocks are aligned
+     play(tx0, p, driveframe1);
+     delay[20dt] driveframe;
```
```suggestion
     delay[20dt] driveframe1;
```

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -290,28 +298,262 @@ pushing the results into some buffer which is then accessed outside the program.
+     frame frame0 = Frame(1.0, 5e9, 0.0);
+  }
+
+  defcal cross_resonance %0 %1 {
```
We've switched to using $q for physical qubits, as `%` refers to the modulo operator. See https://qiskit.github.io/openqasm/language/pulses.html

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -290,28 +298,262 @@ pushing the results into some buffer which is then accessed outside the program.
+      channel d0 = txch(%0, "drive");
+      channel d1 = txch(%1, "drive");
+
+      waveform wf1 = gaussian_square(1., 1024dt, 32dt, 128dt);
```
Not a big deal, but I think these arguments (and their ordering) may differ from your definition given above of `gaussian_square`

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -290,28 +298,262 @@ pushing the results into some buffer which is then accessed outside the program.
+In this simple example, the signal chain is composed of two electro-optic modulators (EOM) and
+an acousto-optic deflector (AOD). The EOMs put sidebands on the laser light while the AOD diffracts
+the light in an amount proportional to the frequency of the RF drive. This example was chosen
+because it is similar in spirit to the work by Levine et al. except that phase control is exerted
```
Link for Levine et al.? I believe it was in the google doc

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -290,28 +298,262 @@ pushing the results into some buffer which is then accessed outside the program.
+            play(aod_channel, q1_π_half_sig, q1_frame);
+
+        // Fence all then time delay
+        fence(.*)
```
`fence`->`barrier`

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -290,28 +298,262 @@ pushing the results into some buffer which is then accessed outside the program.
+    q1_frame = Frame(q1_ro_freq, 0); // time 0
+
+    // flat-top readout waveforms
+    waveform q0_ro_wf = constant(amplitude=0.1, time=...);
```
`constant` should have a `length`, not `time` I would think?

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -290,28 +298,262 @@ pushing the results into some buffer which is then accessed outside the program.
+Incommensurate Rates
+~~~~~~~~~~~~~~~~~~~~
+
+Since the frame can be played on multiple channels, there may be an issue with sample rates.
```
At the top, I believe you said `dt` is tied to the frame. But it seems here it's tied to the channel (which also certainly makes sense...)?

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,234 @@ could even reference multiple qubits in a chain.
 
-Pulses are scheduled using the ``play`` instruction. These instructions may
+Waveforms are scheduled using the ``play`` instruction. These instructions may
 only appear inside a ``defcal`` block!
```
Should we consider allowing `play` (and `capture`) to be done for a specified period of time? That way you can choose the samples you want without having to explicitly shift or clip waveforms.

psivaraj replied:
We _could_.

zachschoenfeld33 commented on source/language/pulses.rst:
```diff
@@ -98,20 +98,18 @@ of prior versions of OpenQASM.
+The contents of ``defcal`` bodies are subject to the restriction they must have a definite length
+known at compile time, regardless of the parameters passed in or the state of the system when
+called. This allows the compiler to properly resolve ``lengthof(...)`` calls and
+allows for optimizations. If there is to be control flow in the ``defcal``, each branch of the
```
I think we should revisit the question of control flow. You could likely do most of it outside of the `defcal` and leave the specific pulse sequence to be defined in the `defcal`. This would then maintain the prior requirement. I don't have a strong opinion either way though

caryan replied:
The main motivator for control flow inside the `defcal` is to be able to `defcal` the `reset` instruction.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -290,28 +298,262 @@ pushing the results into some buffer which is then accessed outside the program.
+    q2_π_half_sig = mix(π_half_sig, sine(q2_π_half_amp, q2_pos_freq-qubit_freq, 0.0, π_half_time));
+    q3_π_half_sig = mix(π_half_sig, sine(q3_π_half_amp, q3_pos_freq-qubit_freq, 0.0, π_half_time));
+
+    for τ in [0: T]:
```
Do we want loops within the `defcal`? You could make the `defcal` the pulse sequence within the loop

psivaraj replied:
Good question. I would say yes so long as they "have a definite length". Otherwise, I'm wondering where the setup information for the `q#_π_half_sig` would go in this example.  Do we allow for nested `defcal` blocks?

zachschoenfeld33 replied:
yeah that's a fair point. It could be made global, but that might be too much. I think this is fine

zachschoenfeld33 reviewed (commented):
Added some general comments. Looks very nice so far (and very similar to many of the ideas in the signal proposal)

psivaraj committed:
CR, round 1.
- Remove named arguments
- Include reference
- Fix waveform argument types
- Typos, odds, and ends

stevenheidel commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,249 @@ could even reference multiple qubits in a chain.
+- ``re(wf: waveform) -> waveform`` - Real component of input waveform.
+- ``im(wf: waveform) -> waveform`` - Imaginary component of input waveform.
+- ``abs(wf: waveform) -> waveform`` - Transform waveform as norm of input. waveform
+- ``phase_shift(wf: waveform, ang: angle) -> waveform`` - Signal with relative phase, ang.
```
Why would I need this vs shifting phase on the frame?

caryan replied:
If you want to sum phase shifted versions of a waveform you would need it.

stevenheidel commented on source/language/openpulse.rst:
```diff
@@ -290,28 +313,261 @@ pushing the results into some buffer which is then accessed outside the program.
+        play(aod_channel, q1_π_half_sig, q3_frame);
+
+        // Time delay all
+        delay(.*, τ/2)
```
Rather than introducing new syntax, does it make sense for a no argument delay/barrier to delay all the frames in scope? That's how the spec is currently written for qubits.

psivaraj replied:
Perfect. Fixed.

stevenheidel commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,249 @@ could even reference multiple qubits in a chain.
+We can manipulate the ``waveform`` types using the following signal processing functions to produce
+new waveforms
+
+- ``mix(wf1: waveform, wf2: waveform)`` -> waveform - Mix two input waveforms to produce a new waveform.
```
mix and scale are the only uses I see in the examples, makes me wonder whether it's worth postponing adding some of these other ones

caryan replied:
Because frames are cheap now we may not need it but I can see folks wanting `sum` and `piecewise`.

taalexander replied:
`phase` may be useful to produce a Y from an X pulse, but this could be equivalently realized with a frame.

psivaraj replied:
> phase may be useful to produce a Y from an X pulse, but this could be equivalently realized with a frame.

That's true, but since that phase is tracking this might still be useful. Or you may want to define a DRAG pulse as as sum of an X and a Y with some scaling.

zachschoenfeld33 commented:
Linking signal proposal (https://github.com/Qiskit/openqasm/pull/178) to this issue for convenient reference

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -1,14 +1,15 @@
 =================
 
-In addition to OpenQASM instructions, ``defcal`` blocks may contain OpenPulse 
+In addition to OpenQASM instructions, ``defcal`` blocks may contain OpenPulse
```
Might be nice to incorporate some of the intro paragraph from the signal proposal. It provides nice high level overview of how OpenPulse fits into OpenQASM. Here is that text:

```
OpenQASM allows users to provide the target system's implementation of quantum operations with
 ``defcal`` `blocks <pulses.rst>`_ . Calibration grammars are open to extension for system
 implementors. In this document, we outline one such grammar, OpenPulse, which maybe selected
 within a supporting compiler through the declaration `defcalgrammar "openpulse" <VERSION>`.
```

psivaraj replied:
Added.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -1,14 +1,15 @@
 appearing inside of a ``defcal`` block.
 
 These instructions are motivated by the original OpenPulse specification which
 was defined as a JSON wire-format for pulse-level quantum programs in the paper
```
```suggestion
, a JSON wire-format for pulse-level quantum programs defined in the paper
```

psivaraj replied:
Fixed.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -1,14 +1,15 @@
-`Qiskit Backend Specifications for OpenQASM and OpenPulse Experiment <https://arxiv.org/abs/1809.03452>`_. 
-The text format described here has several advantages over the equivalent JSON 
+`Qiskit Backend Specifications for OpenQASM and OpenPulse Experiment
+<https://arxiv.org/abs/1809.03452>`_.
```
Should this be a citation? (In line w/ the other citations that were added)

psivaraj replied:
Added.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -1,14 +1,15 @@
 instructions. Certain OpenPulse instructions are further restricted to only
 appearing inside of a ``defcal`` block.
 
 These instructions are motivated by the original OpenPulse specification which
```
```suggestion
This grammar is motivated by the original OpenPulse specification which
```

psivaraj replied:
Re-wrote it take a large chunk of the `Signal` proposal introduction.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -1,14 +1,15 @@
-The text format described here has several advantages over the equivalent JSON 
+`Qiskit Backend Specifications for OpenQASM and OpenPulse Experiment
+<https://arxiv.org/abs/1809.03452>`_.
+The text format described here has several advantages over the equivalent JSON
```
```suggestion
The text format described here has several advantages over the original JSON
```

psivaraj replied:
Fixed.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -1,14 +1,15 @@
+The text format described here has several advantages over the equivalent JSON
 format:
 
 - It is more readable
```
Additions to this list:

```
- Pulse timing is based on instruction ordering and works with programs containing branching control flow (may be the same as the absolute time bullet
- Reusable gate calibrations enable more succinct calibration descriptions.
- Richer ability to compose complex pulses through natural DSP-like operations.
- Clearly defined relationship between pulses, channels and the phases of compilation to hardware resources.
 - Use of multiple oscillators on a single channel at the same time (see :ref:`Geo Gate`)
 ```

psivaraj replied:
Fixed.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -1,14 +1,15 @@
+The text format described here has several advantages over the equivalent JSON
 format:
 
 - It is more readable
```
On the circuit instructions: can we note attachment of physical qubits (which allows microcoding of gate level ops).

psivaraj replied:
Added the intro from the signal proposal.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -1,14 +1,15 @@
+The text format described here has several advantages over the equivalent JSON
 format:
 
 - It is more readable
```
Following this list, can we add the following language (modified from signal proposal):

```
OpenPulse provides a flexible programming model that should extend to many quantum control schemes and hardware. At the core of the OpenPulse grammar are ``pulses``, ``channels`` and ``frames``. ``Pulses`` define a discrete set of samples (or parametric function) which are then played or captured on a ``frame`` on the target device. A ``frame`` is a software abstraction tying an oscillator to the underlying physical ``channel``. This allows 
the programmer to be agnostic to complexities of the device's underlying pulse generation hardware. It is the responsibility of the target device's compiler to map ``pulses`` on ``frames`` to the appropriate instruments in the target hardware.
```

psivaraj replied:
Added.

caryan commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,249 @@ could even reference multiple qubits in a chain.
-``frameof`` command in different places is the same as assigning the frame to a
-constant and referencing it multiple times.
+This will generate a ::math:`pi/2` phase incremented copy of ``driveframe1`` (i.e. with
+the same frequency and time as ``driveframe1``) with the same time as `driveframe1`.
```
repeated "with the same time"

psivaraj replied:
Fixed.

caryan commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,249 @@ could even reference multiple qubits in a chain.
+- Tracking phase by producing a complex value given an input time (i.e. via the mathematical
+  relationship :math:`e^{i\left(2\pi f t + \theta\right)}`,  where `f` is frequency and
+  :math:`\theta` is phase). One motivation for keeping track of phase is to allow pulses to be
+  defined in the rotating frame (i.e. without the carrier) with the effect being
```
What does "without the carrier" mean?

psivaraj replied:
🔥

caryan commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,249 @@ could even reference multiple qubits in a chain.
 
-This has the advantage that one can run any program with any arbitrary frame
-names provided a mapping to the appropriate channels is given.
+Frames are purely virtual and can be constructed using the ``Frame`` command
```
Might not be worth repeating "virtual" as it is discussed in the paragraph just above

psivaraj replied:
:fire:

taalexander commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,249 @@ could even reference multiple qubits in a chain.
-       "{0,1}": "channel2"
-     }
-   }
+When interacting with qubits, it turns out to be quite useful to keep track of a frame of reference,
```
```suggestion
When interacting with qubits, it is powerful to track of a frame of reference,
```

psivaraj replied:
Fixed.

taalexander commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,249 @@ could even reference multiple qubits in a chain.
-     }
-   }
+When interacting with qubits, it turns out to be quite useful to keep track of a frame of reference,
+akin to the rotating frame of a Hamiltonian, throughout the execution of a program. These ``frame``
+types are responsible for two things
```
```suggestion
akin to the rotating frame of a Hamiltonian, throughout the execution of a program. Openpulse provides the ``frame`` type which is responsible for tracking two properties:
```

psivaraj replied:
Fixed.

caryan commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,249 @@ could even reference multiple qubits in a chain.
 only appear inside a ``defcal`` block!
 
-Play instructions have two required parameters:
+Play instructions have three required parameters:
```
suggest ordering the bullets in the same order that they're passed to `play`

psivaraj replied:
Fixed.

caryan commented on source/language/openpulse.rst:
```diff
@@ -261,7 +284,7 @@ process is difficult to describe generically due to the wide variety of
+The only required parameters are the ``channel`` and the ``frame``.
 
 The following are possible parameters that might be included:
 
```
not changed here but typo below: "the measured IQ to distill"

psivaraj replied:
Fixed.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,249 @@ could even reference multiple qubits in a chain.
 hardware provide a method for manipulating those static settings if appropriate.
 
 There are two kinds of channels: transmit channels (sending input to a quantum
 device) and receive channels (reading output from a quantum device).
```
We should include the syntax for channels, in particular the channel accessor syntax.

psivaraj replied:
Added.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,249 @@ could even reference multiple qubits in a chain.
+.. code-block:: javascript
 
-There are two important considersations when dealing with frames.
+  frame driveframe = Frame(5e9, 0.0);
```
Can we add the function signature, ie `Frame(float[size] frequency, angle[size] phase)`?

psivaraj replied:
Added as a comment. Did you want a more explicit code block?

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,249 @@ could even reference multiple qubits in a chain.
 
-.. code-block:: none
+If there are multiple frames associated with a qubit, the map can be extended using a
+``list``
```
Could likely also remove the `list` concept if we go w/ attributes.

psivaraj replied:
🔥

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,249 @@ could even reference multiple qubits in a chain.
+   // amp is waveform amplitude at center
+   // l is the overall length of the waveform
+   // sigma is the standard deviation of waveform
+   gaussian(complex[float[32]]:amp, length:l, length:sigma)
```
Should these be `kernel` functions?

psivaraj replied:
Yes, added.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,249 @@ could even reference multiple qubits in a chain.
+- ``im(wf: waveform) -> waveform`` - Imaginary component of input waveform.
+- ``abs(wf: waveform) -> waveform`` - Transform waveform as norm of input. waveform
+- ``phase_shift(wf: waveform, ang: angle) -> waveform`` - Signal with relative phase, ang.
 
```
Might be nice to consider syntactic sugar like `*` for `mix`, `+` for `sum`, etc...

taalexander reviewed (commented):
Some thoughts on what we should consider before pushing for acceptance:
- [ ] Review from certain internal experimentalists
- [ ] Global setup do we require a configuration section? Are semantics here special?
- [ ] Attributes for the hardware (pulse modifiers, NCO hints, etc.). Are pragmas sufficient?

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -30,227 +31,249 @@ could even reference multiple qubits in a chain.
 
-   // Play a 3 sample pulse on qubit 0's "drive" frame
-   play([1+0*j, 0+1*j, 1/sqrt(2)+1/sqrt(2)*j]) frameof("drive", %0);
+   // Play a 3 sample pulse on the tx0 channel
```
I think adding the general signature would be helpful, ie `play(channel chan, waveform wfm, frame fr)`

psivaraj replied:
Added.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -290,28 +313,261 @@ pushing the results into some buffer which is then accessed outside the program.
-The ``delay`` instruction may take a frame instead of a qubit. The ``barrier`` 
-instruction may also take a list of frames intead of a list of qubits.
+The ``delay`` instruction may take a frame instead of a qubit. The ``barrier``
+instruction may also take a list of frames intead of a list of qubits and aligns the time
```
```suggestion
instruction may also take a list of frames instead of a list of qubits and aligns the time
```

psivaraj replied:
Fixed.

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -290,28 +313,261 @@ pushing the results into some buffer which is then accessed outside the program.
-block.
+``defcal`` blocks have an implicit ``barrier`` on every frame that enters the block,
+meaning that those clocks are guaranteed to be aligned at the start of the block.
+These blocks also need to have a well-defined length, similar to the ``boxas`` block.
```
A `barrier_all` instruction across all frames could be useful.

psivaraj replied:
I've changed it to no argument delaying all frames in scope as per Steven's [comment](https://github.com/stevenheidel/openqasm/pull/2/files?file-filters%5B%5D=.rst#r621802362).

zachschoenfeld33 commented on source/language/openpulse.rst:
```diff
@@ -290,28 +313,261 @@ pushing the results into some buffer which is then accessed outside the program.
+    // now driveframe.time is at 36ns
+  }
+
+This is considered well-defined behavior.
```
We might also wish to add the spectroscopy example from the signal proposal

psivaraj committed:
CR, round 2.

- More complete introduction using aspects from `Signal` proposal
- Defined channel accessors
- More explicit type signatures
- `waveform` templates are kernel functions
- `setframe`/`getframe` and calconfig for global frames and
  parameterized access.
- Filled out more of the `capture` body for variants
- Typos, odds, and ends.

psivaraj committed:
Add citation for Qiskit paper.
